pkg/skaffold/schema/profiles.go

```diff
@@ -34,28 +34,47 @@ import (
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/yamltags"
 )
 
-// ApplyProfiles returns configuration modified by the application
-// of a list of profiles.
-func ApplyProfiles(c *latest.SkaffoldConfig, opts cfg.SkaffoldOptions) error {
+// ApplyProfiles modifies the input skaffold configuration by the application
+// of a list of profiles, and returns the list of applied profiles.
+func ApplyProfiles(c *latest.SkaffoldConfig, opts cfg.SkaffoldOptions, namedProfiles []string) ([]string, error) {
 	byName := profilesByName(c.Profiles)
 
-	profiles, contextSpecificProfiles, err := activatedProfiles(c.Profiles, opts)
+	profiles, contextSpecificProfiles, err := activatedProfiles(c.Profiles, opts, namedProfiles)
 	if err != nil {
-		return fmt.Errorf("finding auto-activated profiles: %w", err)
+		return nil, fmt.Errorf("finding auto-activated profiles: %w", err)
 	}
-
 	for _, name := range profiles {
 		profile, present := byName[name]
 		if !present {
-			return fmt.Errorf("couldn't find profile %s", name)
+			return nil, fmt.Errorf("couldn't find profile %s", name)
 		}
 
 		if err := applyProfile(c, profile); err != nil {
-			return fmt.Errorf("applying profile %q: %w", name, err)
+			return nil, fmt.Errorf("applying profile %q: %w", name, err)
 		}
 	}
 
-	return checkKubeContextConsistency(contextSpecificProfiles, opts.KubeContext, c.Deploy.KubeContext)
+	// remove profiles section for run modes where profiles are already merged into the main pipeline
+	switch opts.Mode() {
+	case cfg.RunModes.Build:
+		fallthrough
+	case cfg.RunModes.Dev:
+		fallthrough
+	case cfg.RunModes.Deploy:
+		fallthrough
+	case cfg.RunModes.Debug:
+		fallthrough
+	case cfg.RunModes.Render:
+		fallthrough
+	case cfg.RunModes.Run:
+		fallthrough
+	case cfg.RunModes.Diagnose:
+		fallthrough
+	case cfg.RunModes.Delete:
+		c.Profiles = nil
+	}
```

Comment:
the current code handles profiles in quite a weird way. In [`profiles.go`](https://github.com/GoogleContainerTools/skaffold/blob/37ed1583747b986c68219a5c8311e72abb4fdcb6/pkg/skaffold/schema/profiles.go#L203) only if at least one profile gets applied then it removes all profile definition objects from the config struct (`cfg.Profiles = nil`). Otherwise it leaves the profile definitions in place. This works out because the `--profile` flag is defined only in a specific selection of run modes (`dev`, `debug`, etc.). So essentially the idea is to retain the profile info if the run mode cares about the config to serialize it as an output (like `generate-pipeline` or `init`).

This switch block makes this expectation explicit.

Reply:
Could this be condensed to a single `case cfg.XX, cfg.YY, cfg.ZZ, ...:`?

Reply:
didn't know `case list` existed. Neat